api/authentication.go

```diff
@@ -61,57 +74,68 @@ func authMiddleware(next http.Handler) http.Handler {
 
 			// check if token is valid
 			if valid && time.Now().Before(validUntil) {
-				// maybe refresh cookie
-				if time.Now().After(validUntil.Add(-cookieRefresh)) {
-					validTokensLock.Lock()
-					validTokens[c.Value] = time.Now()
-					validTokensLock.Unlock()
-				}
+				tracer.Tracef("api: auth token %s is valid, refreshing", c.Value)
+				// refresh cookie
+				validTokensLock.Lock()
+				validTokens[c.Value] = time.Now().Add(cookieTTL)
+				validTokensLock.Unlock()
+				// continue
 				next.ServeHTTP(w, r)
 				return
 			}
-		}
-
-		// get authenticator
-		authFnLock.Lock()
-		authenticator := authFn
-		authFnLock.Unlock()
 
-		// permit if no authenticator set
-		if authenticator == nil {
-			next.ServeHTTP(w, r)
-			return
+			tracer.Tracef("api: provided auth token %s is invalid", c.Value)
 		}
 
 		// get auth decision
-		err = authenticator(server, r)
+		err = authenticator(r.Context(), server, r)
 		if err != nil {
 			if errors.Is(err, ErrAPIAccessDeniedMessage) {
-				log.Warningf("api: denying api access to %s", r.RemoteAddr)
+				tracer.Warningf("api: denying api access to %s", r.RemoteAddr)
 				http.Error(w, err.Error(), http.StatusForbidden)
 			} else {
-				log.Warningf("api: authenticator failed: %s", err)
+				tracer.Warningf("api: authenticator failed: %s", err)
 				http.Error(w, "Internal server error during authentication.", http.StatusInternalServerError)
 			}
 			return
 		}
 
-		// write new cookie
+		// generate new token
 		token, err := rng.Bytes(32) // 256 bit
 		if err != nil {
-			log.Warningf("api: failed to generate random token: %s", err)
+			tracer.Warningf("api: failed to generate random token: %s", err)
 			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
 		}
 		tokenString := base64.RawURLEncoding.EncodeToString(token)
+		// write new cookie
 		http.SetCookie(w, &http.Cookie{
 			Name:     cookieName,
 			Value:    tokenString,
 			HttpOnly: true,
+			SameSite: http.SameSiteStrictMode,
 			MaxAge:   int(cookieTTL.Seconds()),
 		})
+		// save cookie
+		validTokensLock.Lock()
+		validTokens[tokenString] = time.Now().Add(cookieTTL)
+		validTokensLock.Unlock()
```

Comment:
I guess it would be easier to use a JWT here: generate a random key at the beginning and use that key to sign/encrypt JWT issued to clients. This way we can remove the global state validTokens which will make unit testing easier and we also don't need to care about token cleanup. The JWT RFC already defines an "expiresAt" field so we don't need any global state to verify if a token is valid or not.

If we start issuing long-lived API tokens JWT would also be a better choice because we can simply store the signing key instead of tracking each and every token.

Reply:
Created issue to revisit this in the future: https://github.com/safing/portbase/issues/63